src/shared/openssl-util.c

```diff
@@ -624,48 +624,64 @@ int rsa_pkey_to_suitable_key_size(
+/* Generate RSA public key from provided "n" and "e" values. Numbers "n" and "e" must be provided here
+ * in big-endian format, e.g. wrap it with htobe32() for uint32_t. */
 int rsa_pkey_from_n_e(const void *n, size_t n_size, const void *e, size_t e_size, EVP_PKEY **ret) {
         _cleanup_(EVP_PKEY_freep) EVP_PKEY *pkey = NULL;
 
         assert(n);
+        assert(n_size != 0);
         assert(e);
+        assert(e_size != 0);
         assert(ret);
 
-        _cleanup_(EVP_PKEY_CTX_freep) EVP_PKEY_CTX *ctx = EVP_PKEY_CTX_new_id(EVP_PKEY_RSA, NULL);
+#if OPENSSL_VERSION_MAJOR >= 3
+        _cleanup_(EVP_PKEY_CTX_freep) EVP_PKEY_CTX *ctx = EVP_PKEY_CTX_new_from_name(NULL, "RSA", NULL);
         if (!ctx)
                 return log_openssl_errors("Failed to create new EVP_PKEY_CTX");
 
-        _cleanup_(BN_freep) BIGNUM *bn_n = BN_bin2bn(n, n_size, NULL);
-        if (!bn_n)
-                return log_openssl_errors("Failed to create BIGNUM for RSA n");
-
-        _cleanup_(BN_freep) BIGNUM *bn_e = BN_bin2bn(e, e_size, NULL);
-        if (!bn_e)
-                return log_openssl_errors("Failed to create BIGNUM for RSA e");
-
-#if OPENSSL_VERSION_MAJOR >= 3
         if (EVP_PKEY_fromdata_init(ctx) <= 0)
                 return log_openssl_errors("Failed to initialize EVP_PKEY_CTX");
 
-        _cleanup_(OSSL_PARAM_BLD_freep) OSSL_PARAM_BLD *bld = OSSL_PARAM_BLD_new();
-        if (!bld)
-                return log_openssl_errors("Failed to create new OSSL_PARAM_BLD");
+        OSSL_PARAM params[3];
+        _cleanup_free_ void *native_n = NULL, *native_e = NULL;
 
-        if (!OSSL_PARAM_BLD_push_BN(bld, OSSL_PKEY_PARAM_RSA_N, bn_n))
-                return log_openssl_errors("Failed to set RSA OSSL_PKEY_PARAM_RSA_N");
+        uint_fast16_t value = 0x00FF;
+        if (*(uint8_t*)&value == 0) {
+                /* big-endian is native */
+                params[0] = OSSL_PARAM_construct_BN(OSSL_PKEY_PARAM_RSA_N, (void*)n, n_size);
+                params[1] = OSSL_PARAM_construct_BN(OSSL_PKEY_PARAM_RSA_E, (void*)e, e_size);
+        } else {
+                native_n = malloc(n_size);
+                if (!native_n)
+                        return log_oom();
 
-        if (!OSSL_PARAM_BLD_push_BN(bld, OSSL_PKEY_PARAM_RSA_E, bn_e))
-                return log_openssl_errors("Failed to set RSA OSSL_PKEY_PARAM_RSA_E");
+                native_e = malloc(e_size);
+                if (!native_e)
+                        return log_oom();
 
-        _cleanup_(OSSL_PARAM_freep) OSSL_PARAM *params = OSSL_PARAM_BLD_to_param(bld);
-        if (!params)
-                return log_openssl_errors("Failed to build RSA OSSL_PARAM");
+                for (size_t i = 0, k = n_size - 1; i < n_size; i++, k--)
+                        ((uint8_t*) native_n)[i] = ((uint8_t*) n)[k];
+
+                for (size_t i = 0, k = e_size - 1; i < e_size; i++, k--)
+                       ((uint8_t*) native_e)[i] = ((uint8_t*) e)[k];
```

Comment:
please add a helper call for this, that just swaps the bytes of a buffer

Reply:
Is `src/basic/memory-util.{h,c}` the right place for it?
```c++
void *make_swapped(const void *data, size_t size) {
        assert(data);
        assert(size != 0);

        void *p = malloc(size);
        if (p) {
                uint8_t *p_dst = (uint8_t*) p;
                uint8_t *p_src = (uint8_t*) data;
                for (size_t i = 0, k = size - 1; i < size; i++, k--)
                        p_dst[i] = p_src[k];
        }
        return p;
}
```